### PR TITLE
fix(platform-browser): support HTML template content children

### LINKED ISSUE #15557: <template> children not appended to content
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
With `enableLegacyTemplate` set to false (must be hardcoded in `CompilerConfig`, see #15555 for issue in disabling this option), `<template>` child elements are appended using `templateEl.appendChild(childEl)`. This should be `templateEl.content.appendChild(childEl)` to correctly append child elements to the template's document fragment.

screen shot 2017-03-28 at 10 25 29 am

In the above screenshot, the template inside Angular's `<app-root>` is incorrectly rendered. The template outside is rendered correctly in Chrome.

Using `encapsulation: ViewEncapsulation.Native` does not resolve the issue.

**Expected behavior**
If the browser supports `<template>` elements, children should be appended to the template element's content attribute.

**Minimal reproduction of the problem with instructions**
Due to #15555 I cannot provide a Plunker. Create a new project and go to `node_modules/@angular/compiler/@angular/compiler.es5.js` and set `enableLegacyTemplate` to false on line 2287.

```js
    function CompilerConfig(_a) {
        var _b = _a === void 0 ? {} : _a, _c = _b.defaultEncapsulation, defaultEncapsulation = _c === void 0 ? ViewEncapsulation.Emulated : _c, _d = _b.useJit, useJit = _d === void 0 ? true : _d, missingTranslation = _b.missingTranslation, enableLegacyTemplate = _b.enableLegacyTemplate;
        this.defaultEncapsulation = defaultEncapsulation;
        this.useJit = useJit;
        this.missingTranslation = missingTranslation;
        this.enableLegacyTemplate = enableLegacyTemplate !== false;
        // Hard code to false until a valid way to set it is found
        this.enableLegacyTemplate = false;
    }
```

Then, add the following markup to `app.component.html`

```html
<template>
  <div>Hello World</div>
</template>
```

Optionally, add this outside of `<app-root>` in `index.html` to view the correct render. Note that the browser must support HTML template elements for this to work (http://caniuse.com/#feat=template)

**What is the motivation / use case for changing the behavior?**
Compatibility with custom elements

* **Angular version:** 4.0.0

* **Browser:** [all]
 
* **Language:** [all]

* **Node (for AoT issues):** `node --version` =  6.9.5

### PR DESCRIPTION
**Pull request by hotforfeature**
Fixes #15557

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")
```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
#15557

**What is the new behavior?**
`DefaultDomRenderer2` will append, insert, and remove child nodes from a parent's content attribute if the parent's content is a DocumentFragment.

**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: packages/platform-browser/src/dom/dom_renderer.ts [hunk 1]
```diff
@@ -212,6 +214,14 @@ class DefaultDomRenderer2 implements Renderer2 {
     return <() => void>this.eventManager.addEventListener(
                target, event, decoratePreventDefault(callback)) as() => void;
   }
+
+  private nodeOrContent(parent: any): any {
+    if (parent && parent.content instanceof DocumentFragment) {
```
Comment (tbosch) [inline packages/platform-browser/src/dom/dom_renderer.ts]: I think you don't need to check whether `parent` is non null as we check this already before / rely on it being not null.

**Comment by tbosch** [comment]
We need to run this PR through our benchmarks to see the impact of this `instanceof` check... 

In general, we already know during parsing the template that we have a `<template>` element, and could create a way so that we don't need to check during runtime each time. We already have a similar case for shadow dom, see https://github.com/angular/angular/blob/master/packages/platform-browser/src/dom/dom_renderer.ts#L271. I.e. if this causes a performance regression, we could compensate by having a special renderer that supports this case, and we would only use it if there is at least one `<template>` element present.

**Comment by hotforfeature** [comment]
I think those are valid points. My logic in using instanceof was to avoid any potential issue where a custom element that isn't a template has a `content` property. 

Here's a quick jsperf on a couple ideas I had on determining if a node was a template element that supported the content property. https://jsperf.com/template-content-check/

Fastest way would be

```ts
  private nodeOrContent(parent: any): any {
    return parent.content || parent;
  }
```

But it could epically fail if a custom element defined by the user or a third party has a `content` property and children.

I'm not sure if a separate renderer class is the answer though. In https://github.com/angular/angular/blob/master/packages/platform-browser/src/dom/dom_renderer.ts#L70 when the renderer is created, we have to take the view encapsulation in effect.

So what if there's a `<template>` in a Native encapsulated view and one in an Emulated view? We'd need separate `TemplateShadowDomRenderer` and `TemplateEmulatedEncapsulationDomRenderer2` instances. I think it makes more sense to set a flag on `DefaultDomRenderer2` indicating whether or not `<template>` elements are present.

Something like

```ts
class DefaultDomRenderer2 implements Renderer2 {
  constructor(private eventManager: EventManager, private hasTemplate?: boolean) {}

  ...

  private nodeOrContent(parent: any): any {
    if (this.hasTemplate && parent.content instanceof DocumentFragment) {
      return parent.content;
    } else {
      return parent;
    }
  }
}
```

Though `instanceof` is pretty fast imo. This may be unnecessary optimization.

Thoughts?

**Comment by hotforfeature** [comment]
@tbosch any suggestions on my ideas for optimization?

**Comment by tbosch** [comment]
Yes, using a different renderer might not work...

Regarding this PR:
I think the change is good, we need to double check against our benchmarks though.

Thinking about native `<template>` elements a bit more, we also should disable databinding inside of them (similar to `ngNonBindable`):
- e.g. what happens to data binding inside of `<template>`? What if the user code already read the template, cloned it, and then Angular modifies the `<template>`?

**Comment by hotforfeature** [comment]
I agree that databinding inside a `<template>` should be disabled, at least for now.

It would be neat to have Angular propagate bindings to cloned templates, but that seems like quite a big feature to set aside for discussion and a later date.

Though at the moment, using databinding in a template won't "break" the application, it just won't work as the developer expects. I think that's somewhat out of the scope for this PR. We could create a separate ticket to delve deeper into what Angular should be doing inside templates as far as databinding is concerned. From there we can either disable it entirely or create a new feature to support it.

**Comment by tbosch** [comment]
@hotforfeature The point is that using a native `<template>` never worked, as we introduced it as a new feature in 4.x and because of the bug that this PR is trying to solve, nobody was able to use them so far.

If you PR makes them work and we remove the databinding later, this is technically a breaking change, as some people might have relied on it in the mean time.

**Comment by hotforfeature** [comment]
@tbosch any new updates for this PR? I'm not sure what the next step needs to be. 

@hansl mentioned in the issue that this is not a regression. The discussion on whether or not databinding should be disabled in a native `<template>` seems a bit out of scope for what this PR is trying to fix. It's not even possible to use a native `<template>` right now because of #15555.

Do I need to make any changes to the fix?

**Comment by splincode** [comment]
@hotforfeature What is the status of the task? Please closed PR...

**Comment by hotforfeature** [comment]
PR is complete, just been waiting for approval for quite a while.

@tbosch @hansl any update on this? It's getting close to 1 year since the PR was opened.

**Comment by trotyl** [comment]
@mhevery This fix is for the `ProceduralRenderer`, which still plays a part in Ivy, but also require another PR to fix the `ObjectedOrientedRenderer` (inlined instructions).

**Pull request closed without merging**